File: gnunet-build/packages/gnunet/gnunet/files/scheduler.c
/*
 * scheduler.c - gnunet-web version of util/scheduler.c
 * Copyright (C) 2015  David Barksdale <amatus@amatus.name>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "platform.h"
#include "gnunet_util_lib.h"

extern struct GNUNET_SCHEDULER_Task *
GNUNET_SCHEDULER_add_delayed_with_priority_js (double delay,
    enum GNUNET_SCHEDULER_Priority priority,
    GNUNET_SCHEDULER_TaskCallback task,
    void *task_cls);

struct GNUNET_SCHEDULER_Task *
GNUNET_SCHEDULER_add_delayed_with_priority (struct GNUNET_TIME_Relative delay,
    enum GNUNET_SCHEDULER_Priority priority,
    GNUNET_SCHEDULER_TaskCallback task,
    void *task_cls)
{
  if (GNUNET_TIME_UNIT_FOREVER_REL.rel_value_us == delay.rel_value_us) {
    // This is the shutdown task, ignore for now
    return (struct GNUNET_SCHEDULER_Task *)-1;
  }
  return GNUNET_SCHEDULER_add_delayed_with_priority_js(
      (double)(delay.rel_value_us / 1000), priority, task, task_cls);
}

struct GNUNET_SCHEDULER_Task *
GNUNET_SCHEDULER_add_with_priority (enum GNUNET_SCHEDULER_Priority prio,
    GNUNET_SCHEDULER_TaskCallback task,
    void *task_cls)
{
  return GNUNET_SCHEDULER_add_delayed_with_priority (GNUNET_TIME_UNIT_ZERO,
      prio, task, task_cls);
}

struct GNUNET_SCHEDULER_Task *
GNUNET_SCHEDULER_add_delayed (struct GNUNET_TIME_Relative delay,
    GNUNET_SCHEDULER_TaskCallback task, void *task_cls)
{
  return GNUNET_SCHEDULER_add_delayed_with_priority (delay,
      GNUNET_SCHEDULER_PRIORITY_DEFAULT, task, task_cls);
}

struct GNUNET_SCHEDULER_Task *
GNUNET_SCHEDULER_add_now (GNUNET_SCHEDULER_TaskCallback task, void *task_cls)
{
  return GNUNET_SCHEDULER_add_delayed (GNUNET_TIME_UNIT_ZERO, task, task_cls);
}

struct GNUNET_SCHEDULER_Task *
GNUNET_SCHEDULER_add_shutdown (GNUNET_SCHEDULER_TaskCallback task,
                 void *task_cls)
{
  return NULL;
}

void
GNUNET_SCHEDULER_add_continuation (GNUNET_SCHEDULER_TaskCallback task,
    void *task_cls,
    enum GNUNET_SCHEDULER_Reason reason)
{
  GNUNET_SCHEDULER_add_now(task, task_cls);
}

const struct GNUNET_SCHEDULER_TaskContext tc = {
  .reason = GNUNET_SCHEDULER_REASON_WRITE_READY,
  .read_ready = (void *)(intptr_t)1,
  .write_ready = (void *)(intptr_t)2,
};

const struct GNUNET_SCHEDULER_TaskContext *
GNUNET_SCHEDULER_get_task_context ()
{
  return &tc;
}

/* vim: set expandtab ts=2 sw=2: */

File: gnunet-build/packages/gnunet/gnunet/files/scheduler.js
// scheduler.js - scheduler routines for gnunet-web services
// Copyright (C) 2013-2016  David Barksdale <amatus@amat.us>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

mergeInto(LibraryManager.library, {
  $SCHEDULER_TASKS: {},
  GNUNET_SCHEDULER_add_delayed_with_priority_js__deps: ['$SCHEDULER_TASKS'],
  GNUNET_SCHEDULER_add_delayed_with_priority_js:
  function(delay, priority, task, task_cls) {
    //Module.print('GNUNET_SCHEDULER_add_delayed_with_priority(delay='+delay+',pirority='+priority+',task='+task+',task_cls='+task_cls+')');
    var id;
    id = setTimeout(function() {
      delete SCHEDULER_TASKS[id];
      Runtime.dynCall('vi', task, [task_cls]);
    }, delay);
    SCHEDULER_TASKS[id] = {cls: task_cls};
    return id;
  },
  GNUNET_SCHEDULER_add_read_file: function(delay, rfd, task, task_cls) {
    abort();
  },
  GNUNET_SCHEDULER_cancel__deps: ['$SCHEDULER_TASKS'],
  GNUNET_SCHEDULER_cancel: function(task) {
    console.debug("cancelling task", task);
    clearTimeout(task);
    if (task in SCHEDULER_TASKS) {
      var t = SCHEDULER_TASKS[task];
      delete SCHEDULER_TASKS[task];
      if ("socket" in t) {
        if (t.socket in SOCKETS) {
          delete SOCKETS[t.socket]["task"];
        }
      }
      return t.cls;
    }
    return 0;
  },
  GNUNET_SCHEDULER_add_read_net__deps: ['$SOCKETS'],
  GNUNET_SCHEDULER_add_read_net: function(delay, rfd, task, task_cls) {
    console.debug("add_read_net(", delay, rfd, task, task_cls, ")");
    if (!(rfd in SOCKETS)) {
      console.error("socket is not connected?");
      return 0;
    }
    var socket = SOCKETS[rfd];
    if ("task" in socket) {
      console.error("socket already has a read handler");
    }
    var id = setTimeout(function() {
      if (SOCKETS.listening == rfd) {
        if (0 == SOCKETS.incoming.length) {
          return;
        }
      } else if (0 == socket.queue.length) {
        return;
      }
      if (!(id in SCHEDULER_TASKS)) {
        return;
      }
      delete SCHEDULER_TASKS[id];
      delete socket["task"];
      Runtime.dynCall('vi', task, [task_cls]);
    }, 0);
    SCHEDULER_TASKS[id] = {
      cls: task_cls,
      socket: rfd,
    };
    socket["handler"] = task;
    socket["cls"] = task_cls;
    socket["task"] = id;
    console.debug("read task is", id);
    return id;
  },
  GNUNET_SCHEDULER_add_read_net_with_priority__deps: [
    'GNUNET_SCHEDULER_add_read_net'
  ],
  GNUNET_SCHEDULER_add_read_net_with_priority: function(delay, priroity, rfd,
      task, task_cls) {
    return _GNUNET_SCHEDULER_add_read_net(delay, rfd, task, task_cls);
  },
  GNUNET_SCHEDULER_add_write_net__deps: ['$SOCKETS'],
  GNUNET_SCHEDULER_add_write_net: function(delay, wfd, task, task_cls) {
    console.debug("add_write_net(", delay, wfd, task, task_cls, ")");
    if (!(wfd in SOCKETS)) {
      console.error("socket is not connected?");
      return 0;
    }
    // always writable
    var id = setTimeout(function() {
      if (!(id in SCHEDULER_TASKS)) {
        return;
      }
      delete SCHEDULER_TASKS[id];
      Runtime.dynCall('vi', task, [task_cls]);
    }, 0);
    SCHEDULER_TASKS[id] = {cls: task_cls};
    return id;
  },
  GNUNET_SCHEDULER_run: function(task, task_cls) {
    Runtime.dynCall('vi', task, [task_cls]);
    throw 'SimulateInfiniteLoop';
  },
});

// vim: set expandtab ts=2 sw=2:
